### PR TITLE
Switch to new-style pluggy hook wrappers

### LINKED ISSUE #11122: Switch to new-style pluggy hook wrappers
This is a new feature in pluggy 1.1.0: https://pluggy.readthedocs.io/en/latest/changelog.html#pluggy-1-1-0-2023-06-19
    
They are slightly cleaner, faster (if only new-style wrappers are used) and safer (better handling of exceptions thrown after the `yield`). They are fully interoperable with `hookwrapper=True` wrappers.

This will require pluggy>=1.1.0.

### PR DESCRIPTION
**Pull request by bluetech**
Fix #11122.

### CODE REVIEW

**Review by nicoddemus — APPROVED** [review]
Great work!

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Let's wait on https://github.com/pytest-dev/pluggy/issues/405 before moving this forward. :+1:

**Comment by bluetech** [comment]
Updated, pending pluggy 1.2.0 release (the pluggymain variants should pass).

**Comment by bluetech** [comment]
Rebased, should be good to merge.

File: changelog/11122.improvement.rst
Comment (nicoddemus) [inline changelog/11122.improvement.rst:1]: Wonder if this should be considered an improvement? Feels more like an internal change that does not affect users, so perhaps `trivial` would be more appropriate?

**Review by nicoddemus — APPROVED** [review]

File: changelog/11122.improvement.rst
Comment (bluetech) [reply changelog/11122.improvement.rst:1]: It has some behavioral changes in exception handling, however mostly I want plugin authors to see it, so let's keep it as improvement.

**Pull request merged**